extensions/wikia/Captcha/Module/FancyCaptcha.class.php

```diff
@@ -0,0 +1,316 @@
+
+		$resultArr['captcha']['type'] = 'image';
+		$resultArr['captcha']['mime'] = 'image/png';
+		$resultArr['captcha']['id'] = $index;
+		$resultArr['captcha']['url'] = $this->getImageURL( $index );
+	}
+
+	/**
+	 * @param int $index
+	 *
+	 * @return string
+	 */
+	public function getImageURL( $index ) {
+		return \F::app()->wg->Server . '/wikia.php?' . implode( '&', [
+			'controller=Captcha',
+			'method=showImage',
+			'wpCaptchaId=' . urlencode( $index ),
+		] );
+	}
+
+	/**
+	 * Insert the captcha prompt into the edit form.
+	 *
+	 * @param string $class
+	 *
+	 * @return string
+	 * @throws \MWException
+	 */
+	public function getForm( $class = null ) {
+		$info = $this->pickImage();
+		if ( !$info ) {
+			throw new \MWException( "Ran out of captcha images" );
+		}
+
+		// Generate a random key for use of this captcha image in this session.
+		// This is needed so multiple edits in separate tabs or windows can
+		// go through without extra pain.
+		$index = $this->storeCaptcha( $info );
+
+		$this->log( "Captcha id $index using hash ${info['hash']}, salt ${info['salt']}.\n" );
+
+		return "<p>" .
+			\Xml::element( 'img', [
+				'src'    => $this->getImageURL( $index ),
+				'width'  => $info['width'],
+				'height' => $info['height'],
+				'alt'    => '',
+			] ) .
+			"</p>\n" .
+			\Xml::element( 'input', [
+				'type'  => 'hidden',
+				'name'  => 'wpCaptchaId',
+				'id'    => 'wpCaptchaId',
+				'value' => $index,
+			] ) .
+			"<p>" .
+			\Html::element( 'input', [
+				'name' => 'wpCaptchaWord',
+				'id'   => 'wpCaptchaWord',
+				'autocorrect' => 'off',
+				'autocapitalize' => 'off',
+				'required',
+				'class' => $class,
+				'placeholder' => wfMessage( 'captcha-input-placeholder' )->escaped(),
+			] ) .
+			"</p>\n";
+	}
+
+	/**
+	 * Select a previously generated captcha image from the queue.
+	 * @fixme subject to race conditions if lots of files vanish
+	 *
+	 * @return mixed tuple of (salt key, text hash) or false if no image to find
+	 */
+	public function pickImage() {
+		return $this->pickImageDir(
+			\F::app()->wg->CaptchaDirectory,
+			self::DIRECTORY_LEVELS
+		);
+	}
+
+	/**
+	 * @param $directory
+	 * @param $levels
+	 *
+	 * @return array|bool
+	 */
+	public function pickImageDir( $directory, $levels ) {
+		if ( $levels ) {
+			$dirs = [];
+
+			// Check which subdirs are actually present...
+			$dir = opendir( $directory );
+			if ( !$dir ) {
+				return false;
+			}
+			while ( false !== ( $entry = readdir( $dir ) ) ) {
+				if ( ctype_xdigit( $entry ) && strlen( $entry ) == 1 ) {
+					$dirs[] = $entry;
+				}
+			}
+			closedir( $dir );
+
+			$place = mt_rand( 0, count( $dirs ) - 1 );
+			// In case all dirs are not filled,
+			// cycle through next digits...
+			for ( $j = 0; $j < count( $dirs ); $j++ ) {
+				$char = $dirs[( $place + $j ) % count( $dirs )];
+				$return = $this->pickImageDir( "$directory/$char", $levels - 1 );
+				if ( $return ) {
+					return $return;
```

Comment:
Can we use a different variable name here? `return $return;` doesn't tell the reader much what is being returned. Maybe `return $image;`?